### PR TITLE
CosmosClient: Fixes crashing CosmosClient on dispose stage

### LINKED ISSUE #3700: CosmosClient::Dispose throws SynchronizationLockException
**Describe the bug**
`CosmosClient::Dispose()` throws `SynchronizationLockException`

**To Reproduce**
We don't know exactly how to reproduce, but it happens on ASP.NET app shutdown. We have an `IDisposable` Cosmos Client registered to the ASP.NET DI that when disposed by the ASP.NET host, disposes all the different clients it created in a loop:
```cs
foreach (var client in _clients)
{
    client.Value.Value.Dispose();
}
```

Note that `.Value` is `Lazy<T>` so I suspect it could be that some of the clients above are actually created (new `CosmosClient()`) during that loop .

**Expected behavior**
`Dispose()` should not throw exceptions.

**Actual behavior**
`CosmosClient::Dispose()` sometime throws as explained above (see full screenshot below)

**Environment summary**
SDK Version: _Microsoft.Azure.Cosmos_ **3.32.0** (latest at the time of writing)
OS Version: linux (`mcr.microsoft.com/dotnet/aspnet:7.0.2-cbl-mariner2.0-distroless`)

**Additional context**
```
Unhandled exception. System.Threading.SynchronizationLockException: The lock is being disposed while still being used. 
It either is being held by a thread and/or has active waiters waiting to acquire the lock.
   at System.Threading.ReaderWriterLockSlim.Dispose(Boolean disposing)
   at Microsoft.Azure.Documents.Rntbd.LoadBalancingPartition.Dispose()
   at Microsoft.Azure.Documents.Rntbd.LoadBalancingChannel.System.IDisposable.Dispose()
   at Microsoft.Azure.Documents.Rntbd.LoadBalancingChannel.Close()
   at Microsoft.Azure.Documents.Rntbd.ChannelDictionary.Dispose()
   at Microsoft.Azure.Documents.Rntbd.TransportClient.Dispose()
   at Microsoft.Azure.Documents.StoreClientFactory.Dispose()
   at Microsoft.Azure.Cosmos.DocumentClient.Dispose()
   at Microsoft.Azure.Cosmos.ClientContextCore.Dispose(Boolean disposing)
   at Microsoft.Azure.Cosmos.ClientContextCore.Dispose()
   at Microsoft.Azure.Cosmos.CosmosClient.Dispose(Boolean disposing)
   at Microsoft.Azure.Cosmos.CosmosClient.Dispose()
   at Microsoft.Azure.XXX.Documents.Cosmos.CosmosClientFactory.DisposeAsync() in 
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope.DisposeAsync()
--- End of stack trace from previous location ---
   at Microsoft.Extensions.Hosting.Internal.Host.<DisposeAsync>g__DisposeAsync|16_0(Object o)
   at Microsoft.Extensions.Hosting.Internal.Host.DisposeAsync()
   at Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.RunAsync(IHost host, CancellationToken token)
   at Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.Run(IHost host)
   at Microsoft.Azure.XXXProgram.Main(String[] args) 
```

### PR DESCRIPTION
**Pull request by rinatmini**
# Pull Request Template

## Description

Fixes crashing CosmosClient.Dispose if there are any in-flight requests.
We already handle failing HTTPClient during disposing of the CosmosClient if there are any in-flight requests and this PR applies the same for disposing CosmosClient.IStoreClientFactory.
IStoreClientFactory internally closes Microsoft.Azure.Documents.Rntbd.LoadBalancingChannel which in the end also disposes of the own internal [ReaderWriterLockSlim ](https://learn.microsoft.com/en-us/dotnet/api/system.threading.readerwriterlockslim.dispose?view=net-6.0#system-threading-readerwriterlockslim-dispose) and throws SynchronizationLockException if there any of threads are waiting to enter the lock.

## Type of change

- [] Bug fix (non-breaking change which fixes an issue)

## Closing issues

To automatically close an issue: closes #3700

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/DocumentClient.cs [hunk 1]
```diff
@@ -1295,7 +1295,14 @@ public void Dispose()
                 // Dispose only if this store client factory was created and is owned by this instance of document client, otherwise just release the reference
                 if (this.isStoreClientFactoryCreatedInternally)
                 {
-                    this.storeClientFactory.Dispose();
```
Comment (sourabh1007) [inline Microsoft.Azure.Cosmos/src/DocumentClient.cs:1298]: Are you able to replicate this issue? Please add a test.

File: Microsoft.Azure.Cosmos/src/DocumentClient.cs [hunk 2]
```diff
@@ -1295,7 +1295,14 @@ public void Dispose()
...
+                    try
+                    {
+                        this.storeClientFactory.Dispose();
+                    }
+                    catch (Exception exception)
+                    {
+                        DefaultTrace.TraceWarning($"Exception {exception} thrown during dispose of IStoreClientFactory, this could happen if there are inflight request during the dispose of client");
```
Comment (sourabh1007) [inline Microsoft.Azure.Cosmos/src/DocumentClient.cs:1304]: after this exception, should we retry to dispose it again?

File: Microsoft.Azure.Cosmos/src/DocumentClient.cs [hunk 3]
```diff
@@ -1295,7 +1295,14 @@ public void Dispose()
                 // Dispose only if this store client factory was created and is owned by this instance of document client, otherwise just release the reference
                 if (this.isStoreClientFactoryCreatedInternally)
                 {
-                    this.storeClientFactory.Dispose();
+                    try
+                    {
+                        this.storeClientFactory.Dispose();
+                    }
+                    catch (Exception exception)
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/DocumentClient.cs:1302]: If the exception is ObjectDisposedException catch that instead of the generic Exception

**Comment by rinatmini** [comment]
Closed due to fixing in Microsoft.Azure.Cosmos.Direct package on the Microsoft.Azure.Documents.Rntbd.LoadBalancingPartition.Dispose level.

**Pull request closed without merging**